Add bulk get API for state

yaron2 opened the pull request:
This PR does the following:

1. Adds Bulk Get API to HTTP and gRPC
2. Small refactor/logic merge to validation of state store for HTTP and gRPC APIs
3. Updates e2e test app and test scenario

Closes #1338 and #1728

shalabhms commented on pkg/http/api.go:
```diff
@@ -307,18 +313,75 @@ func (a *api) onOutputBindingMessage(reqCtx *fasthttp.RequestCtx) {
+			if err != nil {
+				log.Debugf("bulk get: error getting key %s: %s", k, err)
+			} else if resp != nil && resp.Data != nil {
+				fmt.Println(string(resp.Data))
```
need to remove this print statement.

yaron2 replied:
Done

shalabhms commented on pkg/http/api.go:
```diff
@@ -130,6 +130,12 @@ func (a *api) constructStateEndpoints() []Endpoint {
 		},
+		{
+			Methods: []string{fasthttp.MethodPost},
+			Route:   "state/{storeName}/bulk",
```
Will it route correctly if key happens to be "bulk" ? Could that possibly route to onGetState ?

yaron2 replied:
It does not collide with onGetState because the verb for bulk is POST while the other is GET.

shalabhms replied:
Can we use POST verb for Get semantic ? - as I think POST is for set semantic..

yaron2 replied:
We cannot use GET for this.

shalabhms commented on dapr/proto/runtime/v1/dapr.proto:
```diff
@@ -60,6 +63,29 @@ message GetStateRequest {
 
+// GetBulkStateRequest is the message to get a list of key-value states from specific state store.
+message GetBulkStateRequest {
+  // The name of state store.
```
nit : can remove "The" from comment, though it is there already in other places..if possible might remove this unnecessary "The" .

yaron2 replied:
True, but it's so common on all other comments that I rather not create the inconsistency now..

shalabhms replied:
sounds good.

shalabhms commented on dapr/proto/runtime/v1/dapr.proto:
```diff
@@ -60,6 +63,29 @@ message GetStateRequest {
+  string key = 1;
+
+  // The byte array data
+  bytes data = 2;
```
missing string etag = 3; ? As per comment, it should return etag...

If so, might need to update the comment to include "key" as well...

yaron2 replied:
Addressed in latest commit.

youngbupark commented on pkg/grpc/api.go:
```diff
@@ -212,16 +214,56 @@ func (a *api) InvokeBinding(ctx context.Context, in *runtimev1pb.InvokeBindingRe
+	for _, k := range in.Keys {
+		wg.Add(1)
+
+		go func(k string) {
```
We may have to limit the concurrency here. Otherwise, Some of state store would encounter throttling issue.

yaron2 replied:
How do you think concurrency limiting here should work?

yaron2 replied:
We can optionally add a `Parallelism` field to the request object.

youngbupark replied:
yup adding option makes sense to me.

youngbupark replied:
But we may need to limit 5 or 10 get operation by default.

yaron2 replied:
Great, I'll add the parallelism field.

By default we should probably not limit, that's the default behavior users get in most programming language, C# Parallel.For as an example.

yaron2 replied:
Added.

yaron2 commented:
> I have thought that statestore component exposes bulkget function in [this interface](https://github.com/dapr/components-contrib/blob/8f30f3a6e649801ef85d0c01d5cef1e2e0a006f8/state/store.go#L9-L16), but it is not.
> 
> GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.
> 
> How do you think?
> 
> cc/ @amanbha

We discussed this before, the goal of this is to provide a 100% coverage across all state stores, and so this should not be done on a per component basis.

youngbupark commented:
> > I have thought that statestore component exposes bulkget function in [this interface](https://github.com/dapr/components-contrib/blob/8f30f3a6e649801ef85d0c01d5cef1e2e0a006f8/state/store.go#L9-L16), but it is not.
> > GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> > When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.
> > How do you think?
> > cc/ @amanbha
> 
> We discussed this before, the goal of this is to provide a 100% coverage across all state stores, and so this should not be done on a per component basis.

ok then, can we expand `GetState` instead of introducing new api? The existing `SaveState` api have already supported multiple states.  Making both apis consistent would be better.

> > GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> > When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.

yaron2 commented:
> > > I have thought that statestore component exposes bulkget function in [this interface](https://github.com/dapr/components-contrib/blob/8f30f3a6e649801ef85d0c01d5cef1e2e0a006f8/state/store.go#L9-L16), but it is not.
> > > GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> > > When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.
> > > How do you think?
> > > cc/ @amanbha
> > 
> > 
> > We discussed this before, the goal of this is to provide a 100% coverage across all state stores, and so this should not be done on a per component basis.
> 
> ok then, can we expand `GetState` instead of introducing new api? The existing `SaveState` api have already supported multiple states. Making both apis consistent would be better.
> 
> > > GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> > > When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.

What does it mean to expand `GetState`?

Also, I recommend we don't do this. Currently, if you look at components-contrib, the Bulk Set code is duplicated over all of the components. Ideally there will be a future PR that moves this logic into the Dapr runtime, making it like the implementation in the current PR.

amanbha commented:
> > > > I have thought that statestore component exposes bulkget function in [this interface](https://github.com/dapr/components-contrib/blob/8f30f3a6e649801ef85d0c01d5cef1e2e0a006f8/state/store.go#L9-L16), but it is not.
> > > > GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> > > > When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.
> > > > How do you think?
> > > > cc/ @amanbha
> > > 
> > > 
> > > We discussed this before, the goal of this is to provide a 100% coverage across all state stores, and so this should not be done on a per component basis.
> > 
> > 
> > ok then, can we expand `GetState` instead of introducing new api? The existing `SaveState` api have already supported multiple states. Making both apis consistent would be better.
> > > > GetBulkState API is using the same state get api in GetState. So can we extend the existing `GetState` API to have multiple keys?
> > > > When you look at `SaveState` API, it allows to save multiple keys. To make both APIs consistent and simple, modifying GetState would make sense.
> 
> What does it mean to expand `GetState`?
> 
> Also, I recommend we don't do this. Currently, if you look at components-contrib, the Bulk Set code is duplicated over all of the components. Ideally there will be a future PR that moves this logic into the Dapr runtime, making it like the implementation in the current PR.

How would GetState api look in this case? it will take array of keys and return a array of values?

youngbupark commented:
> Also, I recommend we don't do this. Currently, if you look at components-contrib, the Bulk Set code is duplicated over all of the components. Ideally there will be a future PR that moves this logic into the Dapr runtime, making it like the implementation in the current PR.

We can move bulkset code to runtime side later. it doesn't impact user apps, which is not breaking change.
User would not know how it is implementing it inside component-contrib.

As we focus on user scenario, SaveState supports multiple state with the separate concurrency and consistency. it doesn't guarantee transactional operations.

GetState gets the multiple states separately without any transactional operation.

Providing parity for both get and save would be better.

yaron2 commented:
> > Also, I recommend we don't do this. Currently, if you look at components-contrib, the Bulk Set code is duplicated over all of the components. Ideally there will be a future PR that moves this logic into the Dapr runtime, making it like the implementation in the current PR.
> 
> We can move bulkset code to runtime side later. it doesn't impact user apps, which is not breaking change.
> User would not know how it is implementing it inside component-contrib.
> 
> As we focus on user scenario, SaveState supports multiple state with the separate concurrency and consistency. it doesn't guarantee transactional operations.
> 
> GetState gets the multiple states separately without any transactional operation.
> 
> Providing parity for both get and save would be better.

Correct, this current PR does not validate consistency.

amanbha commented on pkg/http/api.go:
```diff
@@ -307,18 +313,75 @@ func (a *api) onOutputBindingMessage(reqCtx *fasthttp.RequestCtx) {
+	wg.Wait()
+
+	b, _ := a.json.Marshal(bulkResp)
+	respondWithJSON(reqCtx, 200, b)
```
There is no limit on # of keys, what happens when the data to be returned for all keys is greater than the http response message size here, how does user knows about it?

yaron2 replied:
We don't have response message size limits.

amanbha replied:
But there is a limit, its not limitless :)

yaron2 replied:
We don't put message size limits. The same is true for getting a single item that has it's size over any possible message limits. so.. this is not special to bulk of items or a single item.

In both cases, the user will get an error saying the message size was exceeded.

amanbha replied:
This is special to bulk items, as users are more likely to get into this situation with bulk get. There is no good way, dapr can handle this as sizes will be varying across different items and not all components will support paging. So this error seems fine

youngbupark commented:
> How would GetState api look in this case? it will take array of keys and return a array of values?

@amanbha yes. SaveState api gets array of keys.

yaron2 commented:
> > How would GetState api look in this case? it will take array of keys and return a array of values?
> 
> @amanbha yes. SaveState api gets array of keys.

And that's something we will want to change in the future to be consistent with the way we handle multiple items in this PR.

amanbha commented:
> > How would GetState api look in this case? it will take array of keys and return a array of values?
> 
> @amanbha yes. SaveState api gets array of keys.

oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.

yaron2 commented:
> > > How would GetState api look in this case? it will take array of keys and return a array of values?
> > 
> > 
> > @amanbha yes. SaveState api gets array of keys.
> 
> oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.

As discussed, components-contrib should change to remove the BulkSet operation and the Dapr runtime will use just the Set operation on components-contrib. The behavior should be aligned to this PR.

youngbupark commented:
> > > > How would GetState api look in this case? it will take array of keys and return a array of values?
> > > 
> > > 
> > > @amanbha yes. SaveState api gets array of keys.
> > 
> > 
> > oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.
> 
> As discussed, components-contrib should change to remove the BulkSet operation and the Dapr runtime will use just the Set operation on components-contrib. The behavior should be aligned to this PR.

So do you mean we will add SaveBulkState API later?

yaron2 commented:
> > > > > How would GetState api look in this case? it will take array of keys and return a array of values?
> > > > 
> > > > 
> > > > @amanbha yes. SaveState api gets array of keys.
> > > 
> > > 
> > > oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.
> > 
> > 
> > As discussed, components-contrib should change to remove the BulkSet operation and the Dapr runtime will use just the Set operation on components-contrib. The behavior should be aligned to this PR.
> 
> So do you mean we will add SaveBulkState API later?

Yes.

youngbupark commented:
> oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.

@amanbha yes. this will be the breaking change for GetState. may be we need to rename the api from GetState to GetStates.

If possible, I think providing parity of argument for GetXXX and SetXXX would look better. If we want to provide the parity of arguments between Get and Set, we can think about the below options:

1. Separate single state operation and multiple states operation APIs
SaveState <-> GetState : can handle single state 
SaveBulkStates <-> GetBulkStates  : can handle multiple states

2. Providing only multiple state operations.
GetStates <-> SaveStates

youngbupark commented:
> > oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.
> 
> @amanbha yes. this will be the breaking change for GetState. may be we need to rename the api from GetState to GetStates.
> 
> If possible, I think providing parity of argument for GetXXX and SetXXX would look better. If we want to provide the parity of arguments between Get and Set, we can think about the below options:
> 
> 1. Separate single state operation and multiple states operation APIs
>    SaveState <-> GetState : can handle single state
>    SaveBulkStates <-> GetBulkStates  : can handle multiple states
> 2. Providing only multiple state operations.
>    GetStates <-> SaveStates
> 
> > Note : SaveState api already supports multiple states.

@yaron2 do you mean we will go to option 1 eventually ?

> 1. Separate single state operation and multiple states operation APIs
>    SaveState <-> GetState : can handle single state
>    SaveBulkStates <-> GetBulkStates  : can handle multiple states

yaron2 commented:
> > > oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.
> > 
> > 
> > @amanbha yes. this will be the breaking change for GetState. may be we need to rename the api from GetState to GetStates.
> > If possible, I think providing parity of argument for GetXXX and SetXXX would look better. If we want to provide the parity of arguments between Get and Set, we can think about the below options:
> > 
> > 1. Separate single state operation and multiple states operation APIs
> >    SaveState <-> GetState : can handle single state
> >    SaveBulkStates <-> GetBulkStates  : can handle multiple states
> > 2. Providing only multiple state operations.
> >    GetStates <-> SaveStates
> > 
> > > Note : SaveState api already supports multiple states.
> 
> @yaron2 do you mean we will go to option 1 eventually ?
> 
> > 1. Separate single state operation and multiple states operation APIs
> >    SaveState <-> GetState : can handle single state
> >    SaveBulkStates <-> GetBulkStates  : can handle multiple states

Yes, I mean that components-contrib will _only_ have `GetState` and `SaveState`.

amanbha commented:
> > > > oh ok, the return value will also be a map in case of GetState if we add this functionality to GetState, which would adversly affect the usability of GetState for single key.
> > > 
> > > 
> > > @amanbha yes. this will be the breaking change for GetState. may be we need to rename the api from GetState to GetStates.
> > > If possible, I think providing parity of argument for GetXXX and SetXXX would look better. If we want to provide the parity of arguments between Get and Set, we can think about the below options:
> > > 
> > > 1. Separate single state operation and multiple states operation APIs
> > >    SaveState <-> GetState : can handle single state
> > >    SaveBulkStates <-> GetBulkStates  : can handle multiple states
> > > 2. Providing only multiple state operations.
> > >    GetStates <-> SaveStates
> > > 
> > > > Note : SaveState api already supports multiple states.
> > 
> > 
> > @yaron2 do you mean we will go to option 1 eventually ?
> > > 1. Separate single state operation and multiple states operation APIs
> > >    SaveState <-> GetState : can handle single state
> > >    SaveBulkStates <-> GetBulkStates  : can handle multiple states
> 
> Yes, I mean that components-contrib will _only_ have `GetState` and `SaveState`.

Yes I am inclined towards option 1.

youngbupark commented:
> > Yes, I mean that components-contrib will _only_ have `GetState` and `SaveState`.
> 
> Yes I am inclined towards option 1.

@yaron2 @amanbha I created the issue for this. https://github.com/dapr/dapr/issues/1891

youngbupark commented:
/ok-to-test

yaron2 commented:
I didn't run e2e, not sure I got all the parts there. Can you take a look @youngbupark?

I did perform manual tests for both HTTP and gRPC both locally and on k8s.

youngbupark commented:
/ok-to-test

youngbupark commented:
/ok-to-test

youngbupark commented:
/ok-to-test

yaron2 commented:
Updated e2e tests.

As part of these changes, I updated the test scenario validation to account for unordered items.
Today, the test validation assumes that all key/value pairs return in an ordered manner, and this is enforced by the DeepEqual check.

This is now changed to an item-by-item comparison check.

/cc @youngbupark

yaron2 commented:
/ok-to-test

clintsinger commented:
Should this have allowed one to bulk query the Actor states?   I have to get the state from many Actors and it seems that I should be able to do a bulk get on their state using the "<App ID>||<Actor type>||<Actor id>||<state key>" format directly without going through the Actor API.  The intention of this is to reduce the number of network calls needed to query each actor for its individual state.  

When I tried querying the Actor's state (using the .NET SDK), I received an error stating that queries with "||" are not allowed in the key.  This is required for querying the Actor state directly.

Is this a limitation with the bulk state system generally or is specific to the .NET SDK?
